### ISSUE TITLE
CV browser design issues: cvalue_id

### ISSUE THREAD

**Issue by bradfordcondon** [tripal/tripal#297]
# Background

We are creating a CV term browser.  (rather I should stay, stephen has already created a CV term browser).  Ultimately we want to map entities onto the CVterm browser.  The classic case is browsing an ontology (eg Gene Ontology) where counts of features annotated with those terms is in the browser.  The extra challenge is for a given term, we also want to count the features mapped to **children terms**.  Reaching a terminal term allows you to  view the features annotated with that term.

# Specific issue
A solution that lets you count how many fields are using a particular term would be great for other applications and would the proper solution.

Terms can be associated with data in a variety of ways.
* Directly via the bundle type
* The type of the field
* Fields can have nest-layers of values (associative array) where each key is a controlled vocabulary term
* Values of fields can be terms (e.g. feature_cvterm.cvterm_id, xxx_prop.cvalue_id)

It's this last case that is most challenging.

# Proposed Solution

**Create a tripal_cvterm_index table**  with columns:  cvterm_accession, entities array.

* on some periodic basis, iterate through all entities that have fields that provide cvterm associations as values 
* Needs to be updated primarily...
   * When you publish
   * When you edit the cvterm held by a field
      - Targeted update here
   * When you add a new field
      - Look at elementinfo: are you a field thats going to be associating cvterms.
   * When i say so (meaning at the site admin's discretion)
   * Get the value of the cvterm and add an entry to the tripal_cvter_index table.
* For children terms (IE cvtermpath) we can look those up in the cvtermpath table and then go back to the tripal_cvterm_index to look for those entities.

**Comment by bradfordcondon** [comment]
@laceysanderson our group came up with this talking to @spficklin .  We agreed if you had any input it would be super valuable.

**Comment by bradfordcondon** [comment]
hi @laceysanderson - were you able to think about this at all?  Im sure youre busy getting back up to speed with everything.

@spficklin assuming we go ahead with this plan, who is getting this done?  Should abdullah and I work on it, or does your group have plans?

**Comment by laceysanderson** [comment]
@bradfordcondon sorry I'm still drowning in the back-to-work chaos... Feel free to keep reminding/tagging me on specifics you want feedback on and I'll do my best :-)

A bit of clarification:
- tripal_cvterm_index.cvterm_accession = are you thinking this would be a cvterm_id?
- tripal_cvterm_index.entities = would this be a postgres array of entity_ids? Perhaps it would be better as a JSONB field so we could keep track of the way that cvterm connects to the entity. If we make the json object keyed by entity_id and then the value contains connecting info that could be the most helpful.

**Comment by bradfordcondon** [comment]
>tripal_cvterm_index.cvterm_accession = are you thinking this would be a cvterm_id?

yeah!  so at each node of a CV tree (which is a CVterm) you can look up all the children for that node.

>tripal_cvterm_index.entities = would this be a postgres array of entity_ids? Perhaps it would be better as a JSONB field so we could keep track of the way that cvterm connects to the entity. If we make the json object keyed by entity_id and then the value contains connecting info that could be the most helpful.

We were thinking an array... your solution sounds smarter.    Although wouldnt that connecting info be a duplicate of what is in CVterm path?

**Comment by laceysanderson** [comment]
I'm thinking more like tracking how the term connects to each entity: entity type, type field (feature.type_id), additional type of some sort (i.e. some other field), cvterm linking table, key of values array, etc. That way you could break down the count in the cvterm viewer by "strength" of the connection :-) We will need to remember that using JSONB pulls up our required PostgreSQL version to 9.4

With either implementation we might also want a `total count` column that has already done the math. Counting queries are one the slowest in Postgres b/c it needs to [verify rows exist and thus does a sequential scan](https://wiki.postgresql.org/wiki/Slow_Counting).

**Comment by bradfordcondon** [comment]
now that I'm working on this I'm hitting some roadblocks.  We were thinking, create a *field* type that the user can create.  the cvbrowser field would let the user specify which CV to display the tree for, it would automatically only show entities related to the field instance (what im calling the anchor entity, below), and it could be configured to only show entitys of certain bundle types.

Namely, this table is designed to quickly retrieve all of the entities associated iwth a CVterm, but doesn't address the two key features of entity mappings:

* "anchoring" The mapping to a particular entity, for example, Organism A, by only fetching entities related to organism A
* Fetching a subset of entities that belong to a bundle type (only mapping, say, mRNA entities for a particular field)

The second problem could be easily solved via what Lacey is getting at: have the array of entities include bundle type (or even organize by bundle type?)

The first problem is much bigger though.

We have no way of storing entity relationships (I actually proposed this in my nascent days and was told that defining entity relationships  stored outside of Chado is not desirable #167 ).  The relationships are stored in a  Chado linker table, linked to record id's not entity id's.  But more importantly... there's no "index" of these relationships etc, we'd have to populate each ourselves.  And I'm not sure how we would store this information without creating the specialized situations I have in mind (showing biomaterial entites mapped to an analysis, showing genes mapped to an organism).


Any ideas?  It seems to me like each Field would want its own table, with an anchor entity_id, the cvterm, and the array of mappable entities with info allowing easy subsetting based on their entity type.

**Comment by mestato** [comment]
Could you write in support for the most common "anchoring" use cases (organism to feature, analysis to feature, biomaterials to analysis) and then provide the admin a way to specify their own linker relationship?  I'm not sure exactly how this would work in practice (SQL clauses? not pretty but could work), but if we have well documented cases of properly configuring the field with the most common linker table relationships then that should help guide site admins through setting up their own unique one.

Also, if you foresee the field building a table, could you leverage the existing materialized views functionality? Guide a site admin in setting up a table structured to work with this field with whatever anchoring they need, then they can specify the materialized view name to the field?

**Comment by laceysanderson** [comment]
I think you can do this using the original table structure by storing the entity relations as a hierarchical tree in JSONB then querying it using [recursive CTEs](http://tatiyants.com/how-to-navigate-json-trees-in-postgres-using-recursive-ctes/). I’ll try to play with this in the next couple of work days to see if I can get a proof of concept for you.

**Comment by bradfordcondon** [comment]
for what its worth i've got a likely too-slow draft of the backend up at https://github.com/statonlab/tripal/tree/cvbrowser with just a serialized array.  I haven't modified the js tree to plot out the data though.

**Comment by bradfordcondon** [comment]
we created a seperate module for this.  its still got a couple of kinks to work out but its mostly there.  We went with a plain old linker table instead of the array solution.
check it out if you'd like.

https://github.com/statonlab/tripal_cv_xray

**Comment by laceysanderson** [comment]
Looks good :-) Sometimes a simpler solution is better anyway -easier to maintain!

**Comment by spficklin** [comment]
Hey @bradfordcondon what was the reason for going with a separate module?

**Comment by almasaeed2010** [comment]
Hello @spficklin 

Since @bradfordcondon is on vacation I am gonna attempt to answer on his behalf.

My understanding is that when we started developing the module it didn't look generic enough to be part of tripal in that it was not storage agnostic. However, at this stage, I think we arrived at a design that can be genericised further to be part of Tripal.

A few points on the current design that will need to change in order to make this part of tripal:
- Tripal cv browser JS script needs to provide a setting to specify the node url (see https://github.com/tripal/tripal/blob/7.x-3.x/tripal/theme/js/tripal.cv_lookup.js#L46)
- We currently have `cvterm_id` in the linker table along with cv short name and accession but we can drop `cvterm_id` to make the linker generic.
- We need a hook to allow developers to implement their own indexer functionality. We currently only support chado.
- We need a hook to allow developers to generate the count of records associated with each term and entity

Here is a demo of the current implementation: http://160.36.205.61:8095/organism/Fraxinus/excelsior?tripal_pane=group_ontologies

If adding to Tripal directly is the desirable move here, we can start working on putting it in there. Thoughts?

Thanks!

**Comment by bradfordcondon** [comment]
I modularized it because a) feature bloat, b) it required a new table and admin interface to populate the table, c) I wasn't able to make it Chado-agnostic initially.

With Abdullah's help we fixed c somewhat but still, I see it as a small, optional, module.

If you want we can reopen and discuss moving it into core.

Alternatively maybe we discuss moving the **base** cv tree browser you created into the module!
